### PR TITLE
[release/10.0] Add managed FileCatalog MSBuild task

### PR DESCRIPTION
**Pull request by akoeplinger**
Backport of #17098. Adapted to the requirements of release/10.0

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.FileCatalog/CatalogBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+            {
+                if (value is not null && value.Length != ListIdentifierLength)
+                {
+                    throw new ArgumentException(
+                        $"List identifier must be exactly {ListIdentifierLength} bytes.", nameof(value));
+                }
+
+                _listIdentifier = value;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/Microsoft.DotNet.Build.Tasks.FileCatalog/CatalogBuilder.cs:59]: this would require the calling code to screw up, we don't care about that.

File: src/Microsoft.DotNet.Build.Tasks.FileCatalog/GenerateFileCatalog.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                }
+
+                if (!File.Exists(path))
+                {
+                    Log.LogError("File not found: '{0}'.", path);
+                    return false;
+                }
+
+                builder.AddFile(path);
+            }
```
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/Microsoft.DotNet.Build.Tasks.FileCatalog/GenerateFileCatalog.cs:66]: yes but we don't care about this edge case.

File: src/Microsoft.DotNet.Build.Tasks.FileCatalog/GenerateFileCatalog.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+            string? directory = Path.GetDirectoryName(OutputPath);
+            if (!string.IsNullOrEmpty(directory))
+            {
+                Directory.CreateDirectory(directory);
+            }
+
+            builder.WriteTo(OutputPath);
+            Log.LogMessage(MessageImportance.High, "Generated catalog with {0} file(s): {1}", Files.Length, OutputPath);
+            return !Log.HasLoggedErrors;
```
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/Microsoft.DotNet.Build.Tasks.FileCatalog/GenerateFileCatalog.cs:76]: yes but we don't care about this edge case.

File: src/Microsoft.DotNet.Build.Tasks.FileCatalog/CatalogEntry.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        /// <summary>Logical name for the entry (typically the file name). Not embedded in the catalog.</summary>
+        public string Name { get; }
+
+        /// <summary>The raw SHA-1 hash of the content, used as a legacy catalog lookup key.</summary>
+        public byte[] Sha1 { get; }
+
+        /// <summary>The raw SHA-256 hash of the content, recorded in the catalog.</summary>
+        public byte[] Sha256 { get; }
+    }
```
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/Microsoft.DotNet.Build.Tasks.FileCatalog/CatalogEntry.cs:100]: this is fine, if this happens then callers screwed up

File: src/Microsoft.DotNet.Build.Tasks.FileCatalog/CatalogBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+            {
+                sha256.TransformBlock(member.Identifier, 0, member.Identifier.Length, null, 0);
+            }
+
+            sha256.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
+
+            byte[] identifier = new byte[ListIdentifierLength];
+            Array.Copy(sha256.Hash!, identifier, ListIdentifierLength);
+            return identifier;
+        }
```
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/Microsoft.DotNet.Build.Tasks.FileCatalog/CatalogBuilder.cs:249]: I checked this and `TransformBlock` with a null `outputBuffer` does **not** throw on .NET Framework — passing null is the canonical way to feed a hash without an intermediate copy.

Evidence from the .NET Framework reference source (`mscorlib/system/security/cryptography/hashalgorithm.cs`), where the copy is explicitly guarded on `outputBuffer != null` and the up-front validation never touches `outputBuffer`:
```csharp
HashCore(inputBuffer, inputOffset, inputCount);
if ((outputBuffer != null) && ((inputBuffer != outputBuffer) || (inputOffset != outputOffset)))
    Buffer.BlockCopy(inputBuffer, inputOffset, outputBuffer, outputOffset, inputCount);
return inputCount;
```
I also verified empirically under the .NET Framework BCL (via Mono): `TransformBlock(data, 0, data.Length, null, 0)` followed by `TransformFinalBlock` runs fine and produces the correct hash. So there's no cross-TFM correctness issue here and no change is needed.

[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/Microsoft.DotNet.Build.Tasks.FileCatalog/CatalogBuilder.cs:249]: That said, I went ahead and switched `DeriveListIdentifier` to `IncrementalHash` anyway (commit b495bdab5) — not for correctness, but for consistency with the rest of the FileCatalog code (`CatalogEntry.FromFile` and the net472 hash fallback already use it), and to drop the null-`outputBuffer` idiom. Output is byte-for-byte identical (same SHA-256 over the same concatenation), confirmed by the catalog fixture tests.

**Review by ViktorHofer — APPROVED** [review]

File: src/Microsoft.DotNet.Build.Tasks.FileCatalog/GenerateFileCatalog.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        [Required]
+        public ITaskItem[] Files { get; set; } = Array.Empty<ITaskItem>();
+
+        /// <summary>Full path of the .cat file to write.</summary>
+        [Required]
+        public string OutputPath { get; set; } = string.Empty;
+
+        public override bool Execute()
+        {
+            if (string.IsNullOrEmpty(OutputPath))
```
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/Microsoft.DotNet.Build.Tasks.FileCatalog/GenerateFileCatalog.cs:28]: yes but I don't care to handle this edge case

**Review by ViktorHofer — APPROVED** [review]

File: src/Microsoft.DotNet.Build.Tasks.FileCatalog.Tests/CatalogTests.cs [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+            var builder = new CatalogBuilder();
+            for (int i = 0; i < 8; i++)
+            {
+                builder.Add(new CatalogEntry($"f{i}", new byte[] { (byte)i, 0xAB, 0xCD }));
+            }
+
+            List<string> ids = ExtractMembers(builder.Build()).Keys.ToList();
+
+            ids.Should().Equal(ids.OrderBy(id => id, StringComparer.Ordinal));
+        }
```
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/Microsoft.DotNet.Build.Tasks.FileCatalog.Tests/CatalogTests.cs:114]: I don't care about this as it's just a test.
